### ISSUE TITLE
Add streaming support to the minimp3 backend.

### ISSUE THREAD

**Issue by SiegeLord** [liballeg/allegro5#1041]

**Comment by guilt** [comment]
Sidenote: We should consider looking at ffmpeg / avcodec for streaming as well.

**Comment by pmprog** [comment]
Hi @SiegeLord 

I've started some work on this, but I've got a bit stuck.
I load and decode the whole file, and track the position using sample number from the decoded track. It seems to play only about half the file, and then when mp3_stream_update returns a value != buf_size, it crashes the kcm_stream when checking mutex at
https://github.com/pmprog/allegro5/blob/master/addons/audio/kcm_stream.c#L697

My source code is here:
https://github.com/pmprog/allegro5

I haven't tried the actual seeking code, because I was waiting to get the stream to load and play in full, but I can't figure out what's going wrong

**Comment by SiegeLord** [comment]
Hmm, well decoding the whole file into memory seems to defeat the point of doing the streaming, so that seems like a non-starter. I looked at what minimp3 actually supports, and from what it looks like it doesn't actually support streaming... the relevant methods (`mp3dec_ex_read` and `mp3dec_ex_seek`) are stubbed out. The reading one seems not too bad to implement (and contribute back to their library), but seeking might be more tough... maybe I'd maintain a list of frame<->byte offsets.

**Comment by SiegeLord** [comment]
I started implementing the streaming now, just doing a basic version which loads the whole file (without decoding) into memory, and then does basic decoding.

**Comment by pmprog** [comment]
Yes, I noticed that they were stubbed - however, I did come up with an idea, but like most things, I've not really had the time to look into it any further. 

There's this function: `mp3dec_iterate_buf` which goes through checking all the frames. So what I was thinking is that on opening the MP3, make a copy of this function that scans through all the frames in the MP3, and make a note of the file position of each frame, and use that the bases of seeking.

**Comment by fatcerberus** [comment]
@SiegeLord Does 5.2.6 include MP3 support?  It seems like it should, based on the commit history, but it's not mentioned in the changelog so I wanted to make sure.

**Comment by SiegeLord** [comment]
Yep, 5.2.6 has it. Here it is in the changelog: https://github.com/liballeg/allegro5/blob/master/docs/src/changes-5.2.txt#L89

Note that it's unlikely that non-Windows binaries will have it, since MiniMP3 isn't actually packaged in Debian/Ubuntu/OSX. We'd need to figure something extra-ordinary for that to happen.

**Comment by fatcerberus** [comment]
Huh, I did a Ctrl+F in that document for “mp3” and turned up nothing, weird.  Thanks!

I know the pain of not having an available Linux package — miniSphere uses MS ChakraCore as its JavaScript backend and there’s no package for it, forcing me to do this:

https://github.com/fatcerberus/miniSphere/blob/master/Makefile#L126

as part of my build.  That’s not really an option for a library though, we can’t force downstream users to do stuff like that. 😜

**Comment by pmprog** [comment]
Not ideal, but you could use a git submodule. That's how most other projects tend to reference dependency sources on here

**Comment by lieff** [comment]
I've implemented mp3dec_ex_* functions. Now stream with seek to sample should work properly (needs testing).
